### ISSUE TITLE
history --delete not working

### ISSUE THREAD

**Issue by terlar** [fish-shell/fish-shell#3054]
The delete function of history is not working.

### Reproduction Steps:

1. `echo hello`
2. `history echo`
3. `history --delete echo`
4. `history echo`

### Expected behavior:

The entries matching the query `echo` should be deleted from the history.

### Observed behavior:

The entries matching the query `echo` are untouched.

### Additional information:

I have never used this feature before, or at least not recently, so I am not sure when it stopped working. Usually I remove a few entries from the web UI, but when I want to remove a lot of entries that is terribly slow, so that is why I wanted to use the `--delete` feature.

I can delete entries via the web UI.

Video demonstration:
https://asciinema.org/a/18q3wyyxz3ge6hq3xnejhj6oo

---
Fish version: 2.3.0-159-g6594c06

Operating system: Linux 4.5.4-1-ARCH installed through AUR fish-git

Terminal or terminal emulator: termite

**Comment by faho** [comment]
The arguments here are somewhat confusing - `history SOMETHING` searches by prefix, but `history --delete SOMETHING` without an additional `--prefix` or `--contains` will only delete exact matches.

**Comment by terlar** [comment]
Ah, that is extremely useful, I didn't know I could combine those with the `--delete` option. That makes sense. But yeah, could definately be improved in some way, I just assumed it would work the same, just that one was with deleting and one was listing.

Looking in the man page now it is quite clear, not sure how I missed it the first time. I just read the `--delete` part, perhaps mention something about `--prefix` and `--contains` there like we do for `--search`.

**Comment by faho** [comment]
So, how do we want to handle this?

There's already a mention of this in the docs, but it's further down - we could just move it up.

Or we could make "--delete" take a "--exact" option for the current behavior and do "--prefix" without options - like "--search". Or make "--search" do an exact search (which doesn't seem all that useful). Both of course with optional deprecation dance.

@terlar: If this was mentioned in the "--delete" part, you would have understood and it wouldn't have seemed strange?

**Comment by terlar** [comment]
Kind of yes, actually I was using the autocompletions, so `history --<tab>`, so then I saw the `history --delete` with the description `Interactively delete matching history items` and the --search have `Print matching history items, which is the default behavior`. I had also tried the `--search` as well as no argument. So I just assumed it worked the same, then I tried it out and it didn't work, then I checked the man-page quickly but didn't check the bottom part, so didn't see that part explaining about the `--delete`.

I think if the man-page said something like:

```
--delete deletes history items, defaults to exact match, but can be used with --prefix and --contains.
```

Also reading through the docs, the wording of the `--search` options seems a bit weird:
```
--search returns history items in keeping with the --prefix or --contains options.
```

To me that english doesn't make sense, but I am not a native.
